File: src/main/java/com/now/nowbot/service/MessageServiceImpl/AudioService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.OsuApiService.OsuBeatmapApiService;
import com.now.nowbot.throwable.ServiceException.AudioException;
import com.now.nowbot.util.Instructions;
import jakarta.annotation.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.Objects;

@Service("AUDIO")
public class AudioService implements MessageService<AudioService.AudioParam> {
    private static final Logger log = LoggerFactory.getLogger(AudioService.class);

    @Resource
    OsuBeatmapApiService beatmapApiService;
    @Resource
    WebClient osuApiWebClient;

    public static class AudioParam {
        Boolean isBid;
        Integer id;
    }

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<AudioParam> data) throws Exception {
        var matcher = Instructions.AUDIO.matcher(messageText);
        if (!matcher.find()) {
            return false;
        }
        var param = new AudioParam();
        var id_str = matcher.group("id");
        var type = matcher.group("type");

        if (Objects.isNull(id_str)) {
            throw new AudioException(AudioException.Type.SONG_Parameter_NoBid);
        }


        try {
            param.id = Integer.parseInt(id_str);
        } catch (NumberFormatException e) {
            throw new AudioException(AudioException.Type.SONG_Parameter_BidError);
        }

        param.isBid = Objects.equals(type, "b") || Objects.equals(type, "bid");

        data.setValue(param);
        return true;
    }

    @Override
    public void HandleMessage(MessageEvent event, AudioParam param) throws Throwable {
        var from = event.getSubject();

        int sid = 0;

        byte[] voice;

        if (param.isBid) {
            try {
                sid = beatmapApiService.getBeatMapInfo(param.id).getSetID();
            } catch (Exception e) {
                throw new AudioException(AudioException.Type.SONG_Map_NotFound);
            }

            try {
                voice = getVoice(sid);
            } catch (Exception e) {
                log.error("音频下载失败：", e);
                throw new AudioException(AudioException.Type.SONG_Download_Error);
            }

        } else {
            // isSid
            try {
                voice = getVoice(sid);
            } catch (Exception e) {
                //输入的不是 SID
                try {
                    sid = beatmapApiService.getBeatMapInfo(param.id).getSetID();
                } catch (Exception e1) {
                    throw new AudioException(AudioException.Type.SONG_Map_NotFound);
                }

                try {
                    voice = getVoice(sid);
                } catch (Exception e2) {
                    log.error("音频下载失败、附加转换失败：", e2);
                    throw new AudioException(AudioException.Type.SONG_Download_Error);
                }
            }
        }

        try {
            from.sendVoice(voice);
        } catch (Exception e) {
            log.error("音频发送失败：", e);
            throw new AudioException(AudioException.Type.SONG_Send_Error);
        }
    }


    private byte[] getVoice(int sid) {
        var url = STR."https://b.ppy.sh/preview/\{sid}.mp3";

        return osuApiWebClient
                .get()
                .uri(url)
                .retrieve()
                .bodyToMono(byte[].class)
                .block();
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/AudioService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.dao.ServiceCallStatisticsDao
import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.model.enums.IDType
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.messageServiceImpl.AudioService.AudioParam
import com.now.nowbot.service.osuApiService.OsuBeatmapApiService
import com.now.nowbot.throwable.botRuntimeException.IllegalArgumentException
import com.now.nowbot.throwable.botRuntimeException.NoSuchElementException
import com.now.nowbot.util.Instruction
import com.now.nowbot.util.command.FLAG_ID
import com.now.nowbot.util.command.FLAG_TYPE
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import kotlin.math.absoluteValue

@Service("AUDIO")
class AudioService(
    private val beatmapApiService: OsuBeatmapApiService,
    private val dao: ServiceCallStatisticsDao,
) : MessageService<AudioParam> {

    data class AudioParam(var id: Long = 0, var type: IDType = IDType.BeatmapID)

    @Throws(Exception::class)
    override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<AudioParam>,
    ): Boolean {
        val matcher = Instruction.AUDIO.matcher(messageText)
        if (!matcher.find()) {
            return false
        }

        val (inputType, inputID) = IDType.parse(matcher.group(FLAG_TYPE), matcher.group(FLAG_ID))

        val id: Long
        val type: IDType

        if (inputID == null) {
            val last = dao.getLastBeatmapID(event)
            id = last ?: throw IllegalArgumentException.WrongException.Audio()
            type = IDType.BeatmapID
        } else if (inputID > 0) {
            id = inputID
            type = inputType
        } else {
            id = inputID.absoluteValue
            type = inputType
        }

        data.value = AudioParam(id, type)
        return true
    }

    @Throws(Throwable::class)
    override fun handleMessage(event: MessageEvent, param: AudioParam): ServiceCallStatistic? {
        var type: IDType = param.type

        val voice = when (param.type) {
            IDType.BeatmapID -> getVoiceFromBID(param.id) ?: run {
                type = IDType.BeatmapsetID
                getVoiceFromSID(param.id)
            }

            IDType.BeatmapsetID -> getVoiceFromSID(param.id) ?: run {
                type = IDType.BeatmapID
                getVoiceFromBID(param.id)
            }
        }

        if (voice == null) {
            log.info("谱面试听：无法获取 ${type.abbr}${param.id} 的音频")
            throw NoSuchElementException.Audio()
        }

        event.replyVoiceAsync(voice, { e ->
            log.error("谱面试听：发送失败", e)
        })

        return when (type) {
            IDType.BeatmapID -> ServiceCallStatistic.build(event, beatmapID = param.id)
            IDType.BeatmapsetID -> ServiceCallStatistic.build(event, beatmapsetID = param.id)

        }
    }

    private fun getVoiceFromSID(sid: Long): ByteArray? {
        return beatmapApiService.getVoice(sid)
    }

    private fun getVoiceFromBID(bid: Long): ByteArray? {
        val setID = runCatching { beatmapApiService.getBeatmapsetIDFromBeatmapIDByExtend(bid) ?: error("null") }
            .recoverCatching { beatmapApiService.getBeatmapFromDatabase(bid).beatmapsetID }
            .recoverCatching { beatmapApiService.getBeatmap(bid).beatmapsetID }
            .getOrNull() ?: return null

        return beatmapApiService.getVoice(setID)
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(AudioService::class.java)
    }
}
